crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,242 @@
+use crate::price_estimation::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use anyhow::Result;
+use futures::future::WeakShared;
+use futures::FutureExt;
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+
+type EstimationResult = Result<Estimate, PriceEstimationError>;
+type SharedEstimationRequest = WeakShared<Pin<Box<dyn Future<Output = EstimationResult> + Send>>>;
+
+struct Inner {
+    estimator: Box<dyn PriceEstimating>,
+    in_flight_requests: Mutex<HashMap<Query, SharedEstimationRequest>>,
+}
+
+impl Inner {
+    fn collect_garbage(&self) {
+        // TODO: Refactor this to use `HashMap::drain_filter` when it's stable:
+        // https://github.com/rust-lang/rust/issues/59618
+        let mut active_requests = self.in_flight_requests.lock().unwrap();
+        let completed_and_ignored_requests: Vec<_> = active_requests
+            .iter()
+            .filter_map(|(query, handle)| match handle.upgrade() {
+                // future terminated
+                Some(fut) if fut.peek().is_some() => Some(*query),
+                // no strong references left
+                None => Some(*query),
+                // something still depends on the future making progress
+                Some(_) => None,
+            })
+            .collect();
+        for query in &completed_and_ignored_requests {
+            active_requests.remove(query);
+        }
+    }
+}
+
+/// A price estimator which doesn't issue another estimation request while an identical one is
+/// already in-flight.
+pub struct BufferingPriceEstimator {
+    inner: Arc<Inner>,
+}
+
+impl BufferingPriceEstimator {
+    pub fn new(estimator: Box<dyn PriceEstimating>) -> Self {
+        Self {
+            inner: Arc::new(Inner {
+                estimator,
+                in_flight_requests: Mutex::new(Default::default()),
+            }),
+        }
+    }
+
+    async fn estimate_buffered(&self, queries: &[Query]) -> Vec<EstimationResult> {
+        let (active_requests, new_requests) = {
+            let mut in_flight_requests = self.inner.in_flight_requests.lock().unwrap();
+
+            // For each `Query` either get an in-flight request or keep the `Query` to forward it to the
+            // inner price estimator.
+            let (active_requests, remaining_queries): (Vec<_>, Vec<_>) = queries
+                .iter()
+                .map(
+                    |query| match in_flight_requests.get(query).map(WeakShared::upgrade) {
+                        Some(Some(active_request)) => (Some(active_request), None),
+                        _ => (None, Some(*query)),
+                    },
```

Comment:
Yes, I see the issue now. 🤔 
To be honest I would just ignore it until it actually becomes a problem in the future. Right nothing which would end up using the `BufferingPriceEstimator` follows the pattern of poll-store-abandon.

If we were to tackle the issue I suspect that solution 3 would be cleaner to implement than solution 2 (no handling of special cases). And because the tasks we are dealing with basically just wait until an external service responds, polling every future to completion will not incur much overhead compared to canceling.

Reply:
I agree that it is ok to ignore it. I would put a description of this in the file somewhere for future reference.